File: server/wordpress/core-rs/src/rest_api/source.rs
use serde::Deserialize;

use webar_core::http;
use webar_data::ser::Serialize;

use crate::id::{CategoryId, CommentId, MediaId, PageId, PostId, TagId, UserId};

macro_rules! archive {
    (enum $i:ident { $($r:ident = $v:literal,)+ }) => {
        #[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
        pub enum $i {
            $(#[serde(rename=$v)] $r,)+
        }
    };
}

#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
pub enum NodeLeaf {
    #[serde(rename = "node")]
    Node,
}
#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
pub enum NodeChild<C> {
    #[serde(rename = "child")]
    Child(C),
}

#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
pub enum BlogChildNode {
    #[serde(rename = "category")]
    Category(CategoryId, NodeLeaf),
    #[serde(rename = "comment")]
    Comment(CommentId, NodeLeaf),
    #[serde(rename = "media")]
    Media(MediaId, NodeLeaf),
    #[serde(rename = "page")]
    Page(PageId, NodeLeaf),
    #[serde(rename = "post")]
    Post(PostId, NodeLeaf),
    #[serde(rename = "tag")]
    Tag(TagId, NodeLeaf),
}
#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
pub enum ArchiveNode<Addr> {
    #[serde(rename = "blog")]
    Blog(Addr, NodeChild<BlogChildNode>),
    #[serde(rename = "user")]
    User(UserId, NodeLeaf),
}

#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
pub enum EdgeLeaf<E> {
    #[serde(rename = "edge")]
    Edge(E),
}
#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
pub enum EdgeChild<C> {
    #[serde(rename = "child")]
    Child(C),
}
#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
pub enum EdgeBranch<E, C> {
    #[serde(rename = "edge")]
    Edge(E),
    #[serde(rename = "child")]
    Child(C),
}

archive!(
    enum PostEdge {
        Comment = "comment",
    }
);
#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
pub enum BlogChildEdge {
    #[serde(rename = "post")]
    Post(PostId, EdgeLeaf<PostEdge>),
}
archive!(
    enum BlogEdge {
        Category = "category",
        Comment = "comment",
        Media = "media",
        Page = "page",
        Post = "post",
        Tag = "tag",
    }
);

#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
pub enum ArchiveChildEdge<Addr> {
    #[serde(rename = "blog")]
    Blog(Addr, EdgeBranch<BlogEdge, BlogChildEdge>),
}
archive!(
    enum ArchiveSelfEdge {
        User = "user",
    }
);

#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(transparent)]
pub struct ArchiveEdge<Addr>(pub EdgeBranch<ArchiveSelfEdge, ArchiveChildEdge<Addr>>);

pub use http::RequestId;

pub type Request<U> = http::Request<RequestId, U, ()>;
pub type Response<B> = http::Response<(), http::HeaderMap<http::HeaderValue>, B>;

#[derive(Debug, Serialize, Deserialize)]
pub enum ResponseData<Addr, R> {
    #[serde(rename = "node")]
    Node {
        #[serde(rename = "type")]
        ty: ArchiveNode<Addr>,
        response: R,
    },
    #[serde(rename = "edge")]
    Edge {
        #[serde(rename = "type")]
        ty: ArchiveEdge<Addr>,
        full: bool,
        responses: Vec<R>,
    },
}

File: server/wordpress/hs/src/Webar/Server/Wordpress/RestApi/Source.hs
{-# LANGUAGE GeneralisedNewtypeDeriving #-}
{-# LANGUAGE TemplateHaskell #-}

module Webar.Server.Wordpress.RestApi.Source
  ( -- * Archive
    Archive (..),

    -- ** node
    NodeLeaf (..),
    NodeChild (..),
    BlogChildNode (..),
    ArchiveNode (..),

    -- ** edge
    EdgeLeaf (..),
    EdgeChild (..),
    EdgeBranch (..),
    PostEdge (..),
    BlogChildEdge (..),
    BlogEdge (..),
    ArchiveChildEdge (..),
    ArchiveSelfEdge (..),
    ArchiveEdge (..),

    -- * Request
    RequestRecord (..),

    -- ** fetch
    FetchInfo,
    FetchId,

    -- ** http request
    H.RequestId (..),
    Request,
    Response,
    HttpInfo (..),

    -- ** api response
    ResponseData (..),
    ApiResponseInfo (..),

    -- * Snapshot
    SnapshotEdge (..),
    SnapshotType (..),
    NodeData,
    NodeContent (..),
    SetData,
    SetContent (..),
    SnapshotInfo (..),
    NodeInfo,
    EdgeInfo,
  )
where

import Data.Vector (Vector)
import Data.Word (Word32)
import Webar.Blob (BlobId)
import Webar.Data.Cbor (FromCbor, ToCbor)
import Webar.Data.Cbor.TH
import Webar.Data.Json (FromJSON, ToJSON)
import Webar.Data.TH
import qualified Webar.Fetch.Http as F
import qualified Webar.Http as H
import Webar.Object (ObjectId)
import Webar.Server.Wordpress.RestApi.Internal.BlobData (NodeData, SetData)
import Webar.Server.Wordpress.RestApi.Types (ApiInfo)
import Webar.Server.Wordpress.Types
import Webar.Types (Timestamp)

data NodeLeaf = NlNode
  deriving (Show, Eq)

newtype NodeChild c = NcChild c
  deriving (Show, Eq)

data BlogChildNode
  = BcnCategory CategoryId {-# UNPACK #-} NodeLeaf
  | BcnComment CommentId {-# UNPACK #-} NodeLeaf
  | BcnMedia MediaId {-# UNPACK #-} NodeLeaf
  | BcnPage PageId {-# UNPACK #-} NodeLeaf
  | BcnPost PostId {-# UNPACK #-} NodeLeaf
  | BcnTag TagId {-# UNPACK #-} NodeLeaf
  deriving (Show, Eq)

data ArchiveNode
  = AnBlog Address (NodeChild BlogChildNode)
  | AnUser UserId {-# UNPACK #-} NodeLeaf
  deriving (Show, Eq)

$( concat
     <$> traverse
       ( \(n, c) ->
           deriveSumData
             defaultSumOptions {constructorTagModifier = camelTo2 '_' . drop c}
             n
       )
       [ (''NodeLeaf, 2),
         (''NodeChild, 2),
         (''BlogChildNode, 3),
         (''ArchiveNode, 2)
       ]
 )

newtype EdgeLeaf e = ElEdge e
  deriving (Show, Eq)

newtype EdgeChild c = EcChild c
  deriving (Show, Eq)

data EdgeBranch e c = EbEdge e | EbChild c
  deriving (Show, Eq)

data PostEdge = APostComment
  deriving (Show, Eq)

data BlogChildEdge
  = BcePost PostId (EdgeLeaf PostEdge)
  deriving (Show, Eq)

data BlogEdge
  = BeCategory
  | BeComment
  | BeMedia
  | BePage
  | BePost
  | BeTag
  deriving (Show, Eq)

data ArchiveChildEdge
  = AceBlog Address (EdgeBranch BlogEdge BlogChildEdge)
  deriving (Show, Eq)

data ArchiveSelfEdge = AseUser
  deriving (Show, Eq)

$( concat
     <$> traverse
       ( \(n, c) ->
           deriveSumData defaultSumOptions {constructorTagModifier = camelTo2 '_' . drop c} n
       )
       [ (''EdgeLeaf, 2),
         (''EdgeChild, 2),
         (''EdgeBranch, 2),
         (''PostEdge, 5),
         (''BlogChildEdge, 3),
         (''BlogEdge, 2),
         (''ArchiveChildEdge, 3),
         (''ArchiveSelfEdge, 3)
       ]
 )

newtype ArchiveEdge = ArchiveEdge (EdgeBranch ArchiveSelfEdge ArchiveChildEdge)
  deriving (Show, Eq, FromCbor, ToCbor, FromJSON, ToJSON)

data Archive
  = ANode ArchiveNode
  | AEdge ArchiveEdge
  deriving (Show, Eq)

deriveSumData
  defaultSumOptions {constructorTagModifier = camelTo2 '_' . tail}
  ''Archive

data RequestRecord = RrFetch | RrHttpRequest | RrApiResponse
  deriving (Show, Eq)

deriveSumData
  defaultSumOptions {constructorTagModifier = camelTo2 '_' . drop 2}
  ''RequestRecord

type FetchInfo = F.FetchInfo ()

type FetchId = F.FetchId ()

type Request = H.Request H.RequestId ()

type Response = H.Response ()

data HttpInfo = HttpInfo
  { hiFetch :: FetchId,
    hiSeq :: Word32,
    hiResponseIndex :: Word32,
    hiRequest :: Request,
    hiResponse :: Response (BlobId H.JsonBody)
  }
  deriving (Show)

deriveProdCbor
  defaultProductOptions {fieldLabelModifier = camelTo2 '_' . drop 2}
  ''HttpInfo

data ResponseData r
  = RdNode
      { -- | record node id so failed request can be recorded (e.g. deleted node)
        rdnType :: ArchiveNode,
        rdnResponse :: r
      }
  | RdEdge
      { rdeType :: ArchiveEdge,
        rdeFull :: Bool,
        rdeResponses :: Vector r
      }
  deriving (Show)

deriveSumData
  defaultSumOptions
    { sumProduct = defaultProductOptions {fieldLabelModifier = camelTo2 '_' . drop 3},
      constructorTagModifier = camelTo2 '_' . drop 2
    }
  ''ResponseData

data ApiResponseInfo = ApiResponseInfo
  { arFetch :: FetchId,
    arSeq :: Word32,
    arApi :: ApiInfo,
    arTimestamp :: Timestamp,
    arResponse :: ResponseData (ObjectId HttpInfo)
  }
  deriving (Show)

deriveProdData
  defaultProductOptions {fieldLabelModifier = camelTo2 '_' . drop 2}
  ''ApiResponseInfo

data SnapshotEdge = SeSet
  deriving (Show, Eq)

deriveSumData
  defaultSumOptions {constructorTagModifier = camelTo2 '_' . drop 2}
  ''SnapshotEdge

data SnapshotType = StNode | StEdge {-# UNPACK #-} SnapshotEdge | StCollection
  deriving (Show, Eq)

deriveSumData
  defaultSumOptions {constructorTagModifier = camelTo2 '_' . drop 2}
  ''SnapshotType

newtype NodeContent t = NcNormal (BlobId (NodeData t))
  deriving (Show)

deriveSumData
  defaultSumOptions {constructorTagModifier = camelTo2 '_' . drop 2}
  ''NodeContent

data SetContent t = ScNormal
  { scContent :: BlobId (SetData t),
    scFull :: Bool
  }
  deriving (Show)

deriveSumData
  defaultSumOptions
    { constructorTagModifier = camelTo2 '_' . drop 2,
      sumProduct = defaultProductOptions {fieldLabelModifier = camelTo2 '_' . drop 2}
    }
  ''SetContent

data SnapshotInfo c = SnapshotInfo
  { siFetch :: FetchId,
    siApiResponse :: ObjectId ApiResponseInfo,
    siApi :: ApiInfo,
    siTimestamp :: Timestamp,
    siContent :: c
  }
  deriving (Show)

deriveProdData
  defaultProductOptions {fieldLabelModifier = camelTo2 '_' . drop 2}
  ''SnapshotInfo

type NodeInfo t = SnapshotInfo (NodeContent t)

type EdgeInfo t = SnapshotInfo (SetContent t)
